src/finch/tensor.py

```diff
@@ -188,18 +191,37 @@ def _elemwise_op(self, op: str, other: Optional["Tensor"] = None) -> "Tensor":
         if other is None:
             result = jl.broadcast(jl.seval(op), self._obj)
         else:
-            axis_x1, axis_x2 = range(self.ndim, 0, -1), range(other.ndim, 0, -1)
+            if not np.isscalar(other):
+                other = jl.permutedims(other._obj, tuple(range(other.ndim, 0, -1)))
             # inverse swizzle, so `broadcast` appends new dims to the front
             result = jl.broadcast(
-                jl.seval(op),
-                jl.permutedims(self._obj, tuple(axis_x1)),
-                jl.permutedims(other._obj, tuple(axis_x2)),
+                jl.seval(op), jl.permutedims(self._obj, tuple(range(self.ndim, 0, -1))), other
             )
             # swizzle back to the original order
             result = jl.permutedims(result, tuple(range(jl.ndims(result), 0, -1)))
 
         return Tensor(result)
 
+    def __bool__(self):
+        return self._to_scalar(bool)
+
+    def __float__(self):
+        return self._to_scalar(float)
+
+    def __int__(self):
+        return self._to_scalar(int)
+
+    def __index__(self):
+        return self._to_scalar(int)
+
+    def __complex__(self):
+        return self._to_scalar(complex)
+
+    def _to_scalar(self, builtin):
+        if self.size != 1 or self.shape != ():
+            raise ValueError(f"{builtin} can be computed for one-element tensors only.")
```

Comment:
```suggestion
        if self.ndim != 0:
            raise ValueError(f"{builtin} can be computed for scalar tensors only.")
```

Reply:
Instead, can we check that all dimensions, if present, are of size 1?

i.e. `if all(dim == 1 for dim in self.shape):`

Reply:
Similarly, could we compute the scalar value by indexing all dimensions with 1?

Reply:
That would be in contrast to the Array API and NumPy.

Reply:
Right, I think we can stick to Array API standard, and support only 0-D arrays, like the suggested change states:
https://data-apis.org/array-api/latest/API_specification/generated/array_api.array.__bool__.html
https://data-apis.org/array-api/latest/API_specification/generated/array_api.array.__int__.html

NumPy already warns that >0D is deprecated:
```python
In [43]: np.array([0]).__int__()
<ipython-input-43-6afc8c458142>:1: DeprecationWarning: Conversion of an array with ndim > 0 to a scalar is deprecated, and will error in future. Ensure you extract a single element from your array before performing this operation. (Deprecated NumPy 1.25.)
  np.array([0]).__int__()
```

Reply:
sounds good!

Reply:
Done!